api/envoy/extensions/network/dns_resolver/cares/v3/cares_dns_resolver.proto

```diff
@@ -33,6 +33,12 @@ message CaresDnsResolverConfig {
   // resolvers. Defaults to false.
   bool use_resolvers_as_fallback = 3;
 
+  // The resolver will query available network interfaces and determine if there are no available
```

Comment:
This makes sense. I'm curious if you had thought about whether it makes sense to have this be even more specific. E.g. if we end up on an outbound socket that is bound to a specific network interface, should we restrict to just what is routable there? I imagine it gets tricky to infer this.

Reply:
Yep, this is related to what @DavidSchinazi referred to in [his question above](https://github.com/envoyproxy/envoy/pull/19287#discussion_r769104515). It is somewhat tricky to do so in Envoy because we don't have a concept of a preferred network interface that spans dns resolution, connection pools, and sockets. 

We have done work in Envoy Mobile to pass down from the Application layer via platform-specific APIs what the current preferred network is. I also have an outstanding ticket to cache DNS resolution results by network interface. And eventually it all could be glued together to get the more refined solution you are suggesting.

For now, this solution improves conditions albeit in a coarser manner.

Reply:
OK, no objections. For my education, can you explain the motivation for doing this outside Envoy Mobile? Or is this motivated by e2e needs of Envoy Mobile, where you have most of the plumbing down of preferred network,  but are missing the last mile here?

Reply:
Gotcha. This feature is indeed just needed for Envoy Mobile. Eventually we believe a dedicated Android DNS resolver will be needed, similar to the iOS resolver. That will, long term, prevent further "pollution" of the c-ares resolver. However, in the short term, the last couple additions to the c-ares resolver that have been motivated by Envoy Mobile seemed non-invasive enough that we decided to prioritize expediency by adding them to this resolver rather than building a dedicated Android one.